### PR TITLE
Fix and improve tracing for nested AssertionScopes

### LINKED ISSUE #1796: ContainEquivalentOf doesn't pass tracing to the assertion output
`WithTracing` without using a `ITraceWriter` is broken in combination with `ContainEquivalentOf` because the latter [discards](https://github.com/fluentassertions/fluentassertions/blob/6846f7f49c322e561bc25c3d9756e1b7cef50782/Src/FluentAssertions/Collections/GenericCollectionAssertions.cs#L867) the results of the nested call to the `EquivalencyValidator`

### PR DESCRIPTION
**Pull request by dennisdoomen**
Fixes #1796 where tracing doesn't work for nested `AssertionScope`s. Also improves the tracing output a little bit.

**After**
```
 With trace:
  collection
  {
    property collection.Name
    {
      Equivalency was proven by ReferenceEqualityEquivalencyStep
    }
    property collection.Age
    {
      Treating property collection.Age as a value type because System.Int32 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Birthdate
    {
      Treating property collection.Birthdate as a value type because System.DateTime overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Id
    {
      Treating property collection.Id as a value type because System.Int64 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    Equivalency was proven by StructuralEqualityEquivalencyStep
  }
  collection
  {
    property collection.Name
    {
      Equivalency was proven by ReferenceEqualityEquivalencyStep
    }
    property collection.Age
    {
      Treating property collection.Age as a value type because System.Int32 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Birthdate
    {
      Treating property collection.Birthdate as a value type because System.DateTime overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Id
    {
      Treating property collection.Id as a value type because System.Int64 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    Equivalency was proven by StructuralEqualityEquivalencyStep
  }
  collection
  {
    property collection.Name
    {
      Equivalency was proven by StringEqualityEquivalencyStep
    }
    property collection.Age
    {
      Treating property collection.Age as a value type because System.Int32 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Birthdate
    {
      Treating property collection.Birthdate as a value type because System.DateTime overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    property collection.Id
    {
      Treating property collection.Id as a value type because System.Int64 overrides Equals.
      Equivalency was proven by ValueTypeEquivalencyStep
    }
    Equivalency was proven by StructuralEqualityEquivalencyStep
  }
```
**Before**
```
With trace:
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Step ValueTypeEquivalencyStep returned ContinueWithNext
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Age as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Birthdate as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Id as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step StructuralEqualityEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Step ValueTypeEquivalencyStep returned ContinueWithNext
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Age as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Birthdate as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Id as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step StructuralEqualityEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Step ValueTypeEquivalencyStep returned ContinueWithNext
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Age as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Birthdate as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step RunAllUserStepsEquivalencyStep returned ContinueWithNext
  Step AutoConversionStep returned ContinueWithNext
  Step ReferenceEqualityEquivalencyStep returned ContinueWithNext
  Step GenericDictionaryEquivalencyStep returned ContinueWithNext
  Step DataSetEquivalencyStep returned ContinueWithNext
  Step DataTableEquivalencyStep returned ContinueWithNext
  Step DataColumnEquivalencyStep returned ContinueWithNext
  Step DataRelationEquivalencyStep returned ContinueWithNext
  Step DataRowCollectionEquivalencyStep returned ContinueWithNext
  Step DataRowEquivalencyStep returned ContinueWithNext
  Step XDocumentEquivalencyStep returned ContinueWithNext
  Step XElementEquivalencyStep returned ContinueWithNext
  Step XAttributeEquivalencyStep returned ContinueWithNext
  Step ConstraintCollectionEquivalencyStep returned ContinueWithNext
  Step ConstraintEquivalencyStep returned ContinueWithNext
  Step DictionaryEquivalencyStep returned ContinueWithNext
  Step MultiDimensionalArrayEquivalencyStep returned ContinueWithNext
  Step GenericEnumerableEquivalencyStep returned ContinueWithNext
  Step EnumerableEquivalencyStep returned ContinueWithNext
  Step StringEqualityEquivalencyStep returned ContinueWithNext
  Step EnumEqualityStep returned ContinueWithNext
  Treating property collection.Id as a value type because Equals must be used.
  Step ValueTypeEquivalencyStep returned AssertionCompleted
  Step StructuralEqualityEquivalencyStep returned AssertionCompleted

```

### CODE REVIEW

**Comment by jnyrup** [comment]
I'll have a look tomorrow.
How did the included tracing output look before this change?

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 1]
```diff
@@ -62,13 +62,14 @@ private static void UpdateScopeWithReportableContext(AssertionScope scope, Compa
 
         private void RunStepsUntilEquivalencyIsProven(Comparands comparands, IEquivalencyValidationContext context)
         {
+            using var _ = context.Tracer.WriteBlock(node => $"{node.Description}");
```
Comment (jnyrup) [inline Src/FluentAssertions/Equivalency/EquivalencyValidator.cs:65]:
```suggestion
            using var _ = context.Tracer.WriteBlock(node => node.Description);
```

**Review by jnyrup — APPROVED** [review]
Awesome diff.
I did found the current output rather noisy the other day at work, where we constantly had to increase `MaxLines` to get the entire picutre

File: Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.ContainEquivalentOf.cs [hunk 2]
```diff
@@ -195,7 +195,34 @@ public void When_collection_does_contain_equivalent_by_including_single_property
             var item = new Customer { Name = "John", Age = 20 };
 
             // Act / Assert
-            collection.Should().ContainEquivalentOf(item, options => options.Including(x => x.Name));
+            collection.Should().ContainEquivalentOf(item, options => options.Including(x => x.Name).WithTracing());
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Collections/CollectionAssertionSpecs.ContainEquivalentOf.cs]: Oops ;-)

**Pull request merged**